File: BackEnd/src/main/java/com/senai/automotiva/entities/Peca.java
package com.senai.automotiva.entities;

import jakarta.persistence.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "pecas")
public class Peca {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 100)
    private String nome;

    @Column(length = 50)
    private String codigo;

    @Column(length = 300)
    private String descricao;

    @Column(name = "preco_custo", nullable = false, precision = 10, scale = 2)
    private BigDecimal precoCusto;

    @Column(name = "preco_venda", nullable = false, precision = 10, scale = 2)
    private BigDecimal precoVenda;

    @Column(name = "quantidade_estoque", nullable = false)
    private Integer quantidadeEstoque = 0;

    @Column(name = "quantidade_minima")
    private Integer quantidadeMinima = 1;

    @Column(length = 50)
    private String unidade = "UN";

    @Column(length = 50)
    private String fornecedor;

    @Column(nullable = false)
    private Boolean ativo = true;

    @Column(name = "atualizado_em")
    private LocalDateTime atualizadoEm;

    @PrePersist
    @PreUpdate
    public void preUpdate() {
        this.atualizadoEm = LocalDateTime.now();
    }

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public String getNome() { return nome; }
    public void setNome(String nome) { this.nome = nome; }

    public String getCodigo() { return codigo; }
    public void setCodigo(String codigo) { this.codigo = codigo; }

    public String getDescricao() { return descricao; }
    public void setDescricao(String descricao) { this.descricao = descricao; }

    public BigDecimal getPrecoCusto() { return precoCusto; }
    public void setPrecoCusto(BigDecimal precoCusto) { this.precoCusto = precoCusto; }

    public BigDecimal getPrecoVenda() { return precoVenda; }
    public void setPrecoVenda(BigDecimal precoVenda) { this.precoVenda = precoVenda; }

    public Integer getQuantidadeEstoque() { return quantidadeEstoque; }
    public void setQuantidadeEstoque(Integer quantidadeEstoque) { this.quantidadeEstoque = quantidadeEstoque; }

    public Integer getQuantidadeMinima() { return quantidadeMinima; }
    public void setQuantidadeMinima(Integer quantidadeMinima) { this.quantidadeMinima = quantidadeMinima; }

    public String getUnidade() { return unidade; }
    public void setUnidade(String unidade) { this.unidade = unidade; }

    public String getFornecedor() { return fornecedor; }
    public void setFornecedor(String fornecedor) { this.fornecedor = fornecedor; }

    public Boolean getAtivo() { return ativo; }
    public void setAtivo(Boolean ativo) { this.ativo = ativo; }

    public LocalDateTime getAtualizadoEm() { return atualizadoEm; }
    public void setAtualizadoEm(LocalDateTime atualizadoEm) { this.atualizadoEm = atualizadoEm; }

    public boolean estoqueBaixo() {
        return this.quantidadeEstoque <= this.quantidadeMinima;
    }
}

File: FrontEnd/src/@types/Peca.ts
// Espelha com.senai.automotiva.dtos.PecaDTO do back-end.

export type Peca = {
    id: number;
    nome: string;
    codigo?: string;
    descricao?: string;
    precoCusto: number;
    precoVenda: number;
    quantidadeEstoque: number;
    quantidadeMinima: number;
    unidade?: string;
    fornecedor?: string;
    ativo: boolean;
    estoqueBaixo: boolean;
};

// Corresponde a PecaDTO.RequisicaoPeca.
export type PecaInput = {
    nome: string;
    codigo?: string;
    descricao?: string;
    precoCusto: number;
    precoVenda: number;
    quantidadeEstoque?: number;
    quantidadeMinima?: number;
    unidade?: string;
    fornecedor?: string;
};

export type TipoMovimentacaoEstoque = "ENTRADA" | "SAIDA" | "AJUSTE";

// Corresponde a PecaDTO.AjusteEstoque.
export type AjusteEstoqueInput = {
    quantidade: number;
    motivo: string;
    tipoMovimentacao: TipoMovimentacaoEstoque;
};
